File: shared/btagent_shared/types/cloud_hunt.py
"""Cloud Control-Plane Hunter schemas (Phase 6 #117 — connector-independent slice).

Defines the data contracts for:
- :class:`CloudIdentity` — IAM principals (roles, users, service accounts, workload
  identities) with trust-policy metadata enabling transitive STS closure analysis.
- :class:`AgenticWorkload` — AI agent workloads (Bedrock AgentCore, Vertex Agent Engine,
  Cloud Run MCP servers, GKE inference) classified as managed vs. shadow.
- :class:`CloudContainmentProposal` — the Phase-C bullet-2 IAM → IR bridge: inert
  containment proposals (revoke role / freeze access key / detach policy) seeded on
  an Investigation when IAM/STS findings are promoted.

These types are the serialisation layer between:
  * :mod:`btagent_shared.hunt.cloud` (pure detection logic, no network)
  * Future CloudTrail/GuardDuty MCP connectors (deferred, blocked on #100)
  * The #119 HuntFinding triage queue

Design notes:
- Pydantic v2 with ``extra="forbid"`` — any undeclared field is a schema violation.
- StrEnum values are lowercase, consistent with the rest of btagent_shared.types.
- Zero heavy dependencies (no LangChain, no boto3, no network).
- No Alembic migration needed — these are in-flight models only.  If persistence is
  desired, see the "Live-wiring TODO" section in the PR description.
"""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# Cloud-provider and identity-kind enumerations
# ---------------------------------------------------------------------------


class CloudProvider(StrEnum):
    """Supported cloud provider identifiers."""

    AWS = "aws"
    AZURE = "azure"
    GCP = "gcp"


class IdentityKind(StrEnum):
    """IAM principal classification.

    ``ROLE`` — AWS IAM Role / Azure Managed Identity (object, not a human).
    ``USER`` — AWS IAM User / GCP User Account / Azure Entra user.
    ``SERVICE_ACCOUNT`` — GCP Service Account; also used for AWS IAM Users that
        are clearly machine-scoped (detected by naming convention).
    ``WORKLOAD_IDENTITY`` — GCP Workload Identity / Azure Federated Credential;
        an identity that spans K8s ↔ cloud without a long-lived secret.
    """

    ROLE = "role"
    USER = "user"
    SERVICE_ACCOUNT = "service_account"
    WORKLOAD_IDENTITY = "workload_identity"


class AgenticWorkloadKind(StrEnum):
    """Managed vs. shadow classification of AI-agent workload types.

    ``BEDROCK_AGENTCORE`` — AWS Bedrock AgentCore agent; first-party managed.
    ``VERTEX_AGENT_ENGINE`` — Google Vertex AI Agent Engine; first-party managed.
    ``CLOUD_RUN_MCP`` — A Cloud Run service acting as an MCP server; may be
        managed (if governance-tagged) or shadow (if discovered without tags).
    ``GKE_INFERENCE`` — Inference workload on GKE; governance state depends on
        presence of required labels.
    ``UNMANAGED`` — A workload that cannot be classified into the above categories
        yet exhibits agent-like capabilities (e.g., a Lambda with LLM SDK calls
        or an ECS task calling Bedrock without Bedrock AgentCore).
    """

    BEDROCK_AGENTCORE = "bedrock_agentcore"
    VERTEX_AGENT_ENGINE = "vertex_agent_engine"
    CLOUD_RUN_MCP = "cloud_run_mcp"
    GKE_INFERENCE = "gke_inference"
    UNMANAGED = "unmanaged"


# ---------------------------------------------------------------------------
# CloudIdentity
# ---------------------------------------------------------------------------


class CloudIdentity(BaseModel):
    """An IAM principal from any cloud provider, with trust-policy metadata.

    The ``can_be_assumed_by`` list captures the *immediate* trust relationships
    declared in the principal's trust policy; transitive closure (multi-hop
    assume-role chains) is computed by
    :func:`btagent_shared.hunt.cloud.build_trust_graph`.

    ``trust_policy`` is stored verbatim (as parsed JSON-ish dict) so the
    graph builder can re-traverse it without re-fetching from the cloud API.
    Live fetching via CloudTrail/IAM connectors is deferred to #100.
    """

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., description="Unique identifier within the fixture/inventory batch.")
    org_id: str = Field(..., description="Tenant scope.")
    provider: CloudProvider
    kind: IdentityKind
    # For AWS this is the full ARN; for GCP the service-account email;
    # for Azure the object ID or UPN.
    arn_or_id: str = Field(..., min_length=1, max_length=512, description="Provider ARN / ID.")
    display_name: str = Field(default="", max_length=300)
    # Raw trust/delegation policy as a parsed dict.  May be None when the
    # identity has no assume-role trust (e.g. a user with no role assumption).
    trust_policy: dict[str, Any] | None = Field(
        default=None,
        description="IAM trust policy (AWS AssumeRolePolicyDocument / GCP binding conditions).",
    )
    # Immediate trustees — populated by the fixture loader or connector shim.
    # Format: ARN / ID strings matching other CloudIdentity.arn_or_id values
    # within the same inventory batch.
    can_be_assumed_by: list[str] = Field(
        default_factory=list,
        description="Principals that can directly assume/impersonate this identity.",
    )
    # Cross-account flag — True when any trustee is from a different account/project.
    has_cross_account_trust: bool = Field(default=False)
    # Governance state — None means "not yet evaluated".
    governance_tagged: bool | None = Field(
        default=None,
        description="True if all required governance tags are present.",
    )
    last_activity: datetime | None = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    # Free-form enrichment from connector / fixture.
    enrichment: dict[str, Any] = Field(default_factory=dict)


# ---------------------------------------------------------------------------
# AgenticWorkload
# ---------------------------------------------------------------------------


class AgenticWorkload(BaseModel):
    """An AI-agent workload inventory record.

    Classified as *managed* (governance-tagged + known kind) or *shadow*
    (untagged, unknown kind, or discovered outside sanctioned patterns).

    Shadow workloads are flagged with ``is_shadow = True`` and emitted into
    the #119 HuntFinding queue with distinct evidence fields so the downstream
    triage agent can route them to a governance workflow.  The governance
    workflow itself is out of scope for this slice (deferred).

    ``risk_score`` is in [0.0, 1.0]; a simple heuristic:
    - +0.4 for ``is_shadow``
    - +0.3 for excessive IAM permissions (``has_overprivileged_identity``)
    - +0.2 for internet-reachable ingress (``internet_reachable``)
    - +0.1 for absence of ``last_activity``
    The cloud detection logic in :mod:`btagent_shared.hunt.cloud` computes
    this deterministically from the above flags.
    """

    model_config = ConfigDict(extra="forbid")

    id: str = Field(..., description="Unique identifier within the fixture/inventory batch.")
    org_id: str = Field(..., description="Tenant scope.")
    provider: CloudProvider
    kind: AgenticWorkloadKind
    # Resource name / ARN / self-link.
    resource_id: str = Field(..., min_length=1, max_length=512)
    display_name: str = Field(default="", max_length=300)
    # The IAM identity this workload runs as.
    identity_ref: str = Field(
        ...,
        description="ARN/ID of the CloudIdentity this workload runs as.",
    )
    # Governance state.
    governance_tagged: bool = Field(
        default=False,
        description="True when all required governance labels/tags are present.",
    )
    # Shadow detection flags.
    is_shadow: bool = Field(
        default=False,
        description="True if the workload is untagged or outside sanctioned patterns.",
    )
    has_overprivileged_identity: bool = Field(
        default=False,
        description="True if the linked identity has broad/wildcard permissions.",
    )
    internet_reachable: bool = Field(
        default=False,
        description="True if the workload exposes a public ingress endpoint.",
    )
    last_activity: datetime | None = Field(
        default=None,
        description="Timestamp of last invocation; None = never observed or data unavailable.",
    )
    risk_score: float = Field(
        default=0.0,
        ge=0.0,
        le=1.0,
        description="Composite risk in [0, 1] computed by cloud.score_workload_risk().",
    )
    # Free-form enrichment from connector / fixture.
    enrichment: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime = Field(default_factory=datetime.utcnow)


# ---------------------------------------------------------------------------
# Containment proposals (#117 Phase C bullet 2 — IAM/STS finding → IR)
# ---------------------------------------------------------------------------


class CloudContainmentActionType(StrEnum):
    """The three cloud control-plane containment verbs this slice proposes.

    ``REVOKE_ROLE`` — revoke the role's active STS sessions / strip the abused
        trust relationship so the assume-role pivot stops working.
    ``FREEZE_ACCESS_KEY`` — deactivate a long-lived access key an attacker
        minted for persistence.
    ``DETACH_POLICY`` — remove an inline/attached policy that granted the
        unexpected privilege.

    These names are *also* the ``action_type`` handed to the #106 containment
    execute service, so they are what shows up in the audit ledger
    (``execute:revoke_role`` etc.).
    """

    REVOKE_ROLE = "revoke_role"
    FREEZE_ACCESS_KEY = "freeze_access_key"
    DETACH_POLICY = "detach_policy"


class CloudContainmentActionStatus(StrEnum):
    """Per-action lifecycle.

    An action is ``PROPOSED`` (inert) until an incident commander accepts the
    proposal; the #106 execute path then flips it to ``EXECUTED`` or ``DENIED``
    (RBAC/approval/safelist refusal — always with an audit row behind it).
    """

    PROPOSED = "proposed"
    EXECUTED = "executed"
    DENIED = "denied"


class CloudContainmentProposalStatus(StrEnum):
    """Lifecycle of the proposal attached to an investigation."""

    PROPOSED = "proposed"
    ACCEPTED = "accepted"
    REJECTED = "rejected"


class CloudContainmentAction(BaseModel):
    """One proposed containment action against one cloud IAM principal.

    Deliberately *inert data*: it carries everything the #106 containment
    execute path needs (``action_type`` / ``connector`` / ``target``) but no
    execution machinery of its own. Nothing in this model can dispatch; the
    only way it becomes an action is a human accepting the proposal through the
    double-gated containment route.
    """

    model_config = ConfigDict(extra="forbid")

    # Stable within the proposal (``cca_1``, ``cca_2``, …) so a partial accept
    # can name exactly which actions to run.
    id: str = Field(..., min_length=1, max_length=64)
    action_type: CloudContainmentActionType
    provider: CloudProvider
    # The principal acted on — an ARN / service-account email / object id. This
    # is the value screened against the org never-touch safelist before any
    # dispatch.
    target: str = Field(..., min_length=1, max_length=512)
    # Connector that would enforce it (``aws_iam`` / ``gcp_iam`` / ``azure_iam``).
    # Live cloud control-plane connectors are deferred to #100, so in practice
    # this dispatches mock-first and fails closed in live mode.
    connector: str = Field(..., min_length=1, max_length=64)
    description: str = Field(default="", max_length=2048)
    # Action-shaped detail: policy_name, user_name, external trustees, …
    parameters: dict[str, Any] = Field(default_factory=dict)
    source_finding_ids: list[str] = Field(default_factory=list)
    status: CloudContainmentActionStatus = CloudContainmentActionStatus.PROPOSED
    # Set once the action has been through the #106 execute path.
    outcome: str = Field(default="", max_length=64)
    audit_id: str | None = None
    message: str = Field(default="", max_length=2048)


class CloudContainmentProposal(BaseModel):
    """Inert containment proposals seeded on promotion of IAM/STS findings.

    Mirrors the identity-hunt :class:`~btagent_shared.types.identity_hunt.RevocationProposal`
    pattern: promotion attaches proposal-shaped data, and a *separate* human
    decision is what can ever make it act. Here that decision is routed through
    the #106 containment execute service, so acceptance inherits the
    ``containment:execute`` RBAC scope, the explicit approved-flag second gate,
    mock-by-default dispatch, the org safelist screen, and an audit row on
    every execute AND every denial.
    """

    model_config = ConfigDict(extra="forbid")

    actions: list[CloudContainmentAction]
    rationale: str = Field(default="", max_length=8192)
    status: CloudContainmentProposalStatus = CloudContainmentProposalStatus.PROPOSED
    decided_by: str | None = None
    decided_at: datetime | None = None
    decision_rationale: str = Field(default="", max_length=8192)

File: frontend/src/types/cloud_hunt.ts
/**
 * Cloud Control-Plane Hunter TypeScript types (#117 Phase B).
 *
 * Mirrors the Python schemas in ``shared/btagent_shared/types/cloud_hunt.py``.
 * String-literal unions match the backend StrEnum values exactly (lowercase).
 *
 * These types are used by:
 *  - ``@/api/cloud``        — API wrapper pre-filtered to domain=cloud
 *  - ``@/stores/cloudStore`` — Zustand store for cloud hunt findings
 *  - ``@/components/cloud/CloudHuntsPage`` — the main page component
 */

// ---------------------------------------------------------------------------
// Enums (mirroring CloudProvider, IdentityKind, AgenticWorkloadKind)
// ---------------------------------------------------------------------------

export type CloudProvider = "aws" | "azure" | "gcp";

export type IdentityKind = "role" | "user" | "service_account" | "workload_identity";

export type AgenticWorkloadKind =
  | "bedrock_agentcore"
  | "vertex_agent_engine"
  | "cloud_run_mcp"
  | "gke_inference"
  | "unmanaged";

// ---------------------------------------------------------------------------
// CloudIdentity — IAM principal with trust-policy metadata
// ---------------------------------------------------------------------------

export interface CloudIdentity {
  id: string;
  org_id: string;
  provider: CloudProvider;
  kind: IdentityKind;
  /** Full ARN (AWS), service-account email (GCP), or object-ID/UPN (Azure). */
  arn_or_id: string;
  display_name: string;
  /** Raw trust / delegation policy; null when the identity has no assume-role trust. */
  trust_policy: Record<string, unknown> | null;
  /**
   * Principals that can directly assume / impersonate this identity.
   * Format: ARN/ID strings matching other ``CloudIdentity.arn_or_id`` values.
   */
  can_be_assumed_by: string[];
  has_cross_account_trust: boolean;
  governance_tagged: boolean | null;
  last_activity: string | null;
  created_at: string;
  enrichment: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// AgenticWorkload — AI-agent workload inventory record
// ---------------------------------------------------------------------------

export interface AgenticWorkload {
  id: string;
  org_id: string;
  provider: CloudProvider;
  kind: AgenticWorkloadKind;
  resource_id: string;
  display_name: string;
  identity_ref: string;
  governance_tagged: boolean;
  is_shadow: boolean;
  has_overprivileged_identity: boolean;
  internet_reachable: boolean;
  last_activity: string | null;
  risk_score: number;
  enrichment: Record<string, unknown>;
  created_at: string;
}

// ---------------------------------------------------------------------------
// Evidence shapes embedded in HuntFinding.evidence for domain=cloud findings
// ---------------------------------------------------------------------------

/**
 * Evidence fields on a cloud HuntFinding.
 *
 * These are extracted from ``HuntFinding.evidence`` (typed as
 * ``Record<string, unknown>``). All fields are optional because the backend
 * may not set every field for every finding.
 */
export interface CloudFindingEvidence {
  /** Cloud provider (aws / azure / gcp). */
  provider?: CloudProvider;
  /** Cloud account / project / subscription ID. */
  account_id?: string;
  /** IAM actor ARN/ID who performed the action. */
  actor_arn?: string;
  /** Target resource ARN/ID. */
  target_arn?: string;
  /**
   * STS assume-role chain trace from the Phase-A ``detect_sts_chaining``
   * detector. Real emitter writes ``path`` (with ``detection==="sts_chaining"``);
   * ``assume_chain`` is accepted as a legacy alias.
   */
  path?: string[];
  /** Legacy alias for ``path`` — older fixtures may still use this key. */
  assume_chain?: string[];
  /** Detection-type tag from the Phase-A emitters (e.g. ``"sts_chaining"``). */
  detection?: string;
  /** Raw trust policy extract (AWS AssumeRolePolicyDocument). */
  trust_policy?: Record<string, unknown>;
  /** Risk score in [0, 1] (only for shadow-workload findings). */
  risk_score?: number;
  /** True when this finding relates to a shadow (unmanaged) agentic workload. */
  shadow_workload?: boolean;
  /**
   * Workload kind from the real ``detect_shadow_workloads`` /
   * ``detect_overprivileged_workload_identity`` emitters (``evidence.kind``).
   * ``workload_kind`` is accepted as a legacy alias.
   */
  kind?: AgenticWorkloadKind;
  workload_kind?: AgenticWorkloadKind;
  /** MITRE technique family used for grouping in the tamper tab. */
  technique_family?: string;
}

// ---------------------------------------------------------------------------
// UI-only aggregated types
// ---------------------------------------------------------------------------

/**
 * One row in the control-plane event timeline.
 * Derived from a HuntFinding with domain=cloud.
 */
export interface CloudTimelineEntry {
  finding_id: string;
  created_at: string;
  provider: CloudProvider;
  account_id: string;
  actor: string;
  target: string;
  technique_ids: string[];
  title: string;
  severity: string;
}

/**
 * One row in the IAM role-graph view.
 * Derived from assume_chain / trust_policy evidence on cloud findings.
 */
export interface IAMRelationship {
  /** The role being assumed (target). */
  source_role: string;
  /** The principal that can assume ``source_role``. */
  trustee: string;
  /** Source finding that produced this relationship. */
  finding_id: string;
  /** Cross-account if trustee is in a different account than source_role. */
  is_cross_account: boolean;
}

/**
 * One cell in the agentic-workload inventory matrix.
 * Rows = provider, columns = workload kind.
 */
export interface WorkloadMatrixCell {
  provider: CloudProvider;
  kind: AgenticWorkloadKind;
  managed_count: number;
  shadow_count: number;
}

/** Active tab in the Cloud Hunts page. */
export type CloudTab = "timeline" | "iam" | "shadow_workloads" | "tamper";

/** Display labels for provider enum values. */
export const CLOUD_PROVIDER_LABELS: Record<CloudProvider, string> = {
  aws: "AWS",
  azure: "Azure",
  gcp: "GCP",
};

/** Display labels for workload kind enum values. */
export const WORKLOAD_KIND_LABELS: Record<AgenticWorkloadKind, string> = {
  bedrock_agentcore: "Bedrock AgentCore",
  vertex_agent_engine: "Vertex Agent Engine",
  cloud_run_mcp: "Cloud Run MCP",
  gke_inference: "GKE Inference",
  unmanaged: "Unmanaged",
};

/** All workload kinds in a stable column order for the matrix table. */
export const WORKLOAD_KINDS_ORDERED: AgenticWorkloadKind[] = [
  "bedrock_agentcore",
  "vertex_agent_engine",
  "cloud_run_mcp",
  "gke_inference",
  "unmanaged",
];

/** All cloud providers in a stable row order for the matrix table. */
export const CLOUD_PROVIDERS_ORDERED: CloudProvider[] = ["aws", "azure", "gcp"];

// ---------------------------------------------------------------------------
// Containment proposals (#117 Phase C bullet 2 — IAM/STS finding → IR)
// ---------------------------------------------------------------------------
//
// Mirrors ``CloudContainmentAction`` / ``CloudContainmentProposal`` in
// ``shared/btagent_shared/types/cloud_hunt.py``. The proposal is *inert data*
// hanging off an Investigation: nothing in it can dispatch. The only thing that
// can make it act is a human accepting it through
// ``POST /cloud/investigations/{id}/containment-proposal/accept``, which routes
// every action through the #106 containment execute service and inherits that
// path's ``containment:execute`` RBAC scope, its explicit approved-flag second
// gate, mock-by-default dispatch, the org never-touch principal safelist, and an
// audit row on every execute AND every denial.

/** The three cloud control-plane containment verbs this slice proposes. */
export type CloudContainmentActionType =
  | "revoke_role"
  | "freeze_access_key"
  | "detach_policy";

/**
 * Per-action lifecycle.
 *
 * ``proposed`` is inert. ``executed`` and ``denied`` are both written by the
 * #106 execute path — and ``denied`` always has a hash-chained audit row behind
 * it, which is why the UI renders it as a recorded guardrail outcome and never
 * as a failed request.
 */
export type CloudContainmentActionStatus = "proposed" | "executed" | "denied";

/** Lifecycle of the proposal attached to an investigation. */
export type CloudContainmentProposalStatus = "proposed" | "accepted" | "rejected";

/** One proposed containment action against one cloud IAM principal. */
export interface CloudContainmentAction {
  /** Stable within the proposal (``cca_1``, ``cca_2``, …) — names a partial accept. */
  id: string;
  action_type: CloudContainmentActionType;
  provider: CloudProvider;
  /** The principal acted on; the value screened against the org safelist. */
  target: string;
  /** Connector that would enforce it (``aws_iam`` / ``gcp_iam`` / ``azure_iam``). */
  connector: string;
  description: string;
  /** Action-shaped evidence detail: reason, event_name, policy_name, trustees, … */
  parameters: Record<string, unknown>;
  source_finding_ids: string[];
  status: CloudContainmentActionStatus;
  /** Set once the action has been through the #106 execute path. */
  outcome: string;
  audit_id: string | null;
  /** Verbatim server reason — the safelist refusal text arrives here. */
  message: string;
}

/** Inert containment proposals seeded on promotion of IAM/STS findings. */
export interface CloudContainmentProposal {
  actions: CloudContainmentAction[];
  rationale: string;
  status: CloudContainmentProposalStatus;
  decided_by: string | null;
  decided_at: string | null;
  decision_rationale: string;
}

/**
 * Accept/reject body — ``approved`` is the explicit HITL half of the gate.
 *
 * Optional here because that is the wire contract (the backend defaults it to
 * ``false``, so an omitted flag can never execute). Callers that *are* accepting
 * should require it locally rather than letting it default — see
 * ``@/api/cloudContainment``.
 */
export interface CloudContainmentDecisionRequest {
  approved?: boolean;
  rationale?: string;
  action_ids?: string[];
}

/** Display labels for the containment verbs. */
export const CLOUD_CONTAINMENT_ACTION_LABELS: Record<
  CloudContainmentActionType,
  string
> = {
  revoke_role: "Revoke role sessions",
  freeze_access_key: "Freeze access key",
  detach_policy: "Detach policy",
};
